### PR TITLE
Disallow exiting from the game if un-cancellable work is active

### PR DESCRIPTION
**Pull request by frenzibyte**
- [ ] Depends on #9034 
- [ ] Depends on #9117 (for testing)

Separated for reviewing ease.

### CODE REVIEW

File: osu.Game.Tests/Visual/Menus/TestSceneSafeExiting.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using osu.Game.Input.Bindings;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+using osu.Game.Screens.Menu;
+using osu.Game.Screens.Menu.Exiting;
+using osu.Game.Tests.Visual.Navigation;
+
+namespace osu.Game.Tests.Visual.Menus
+{
+    [System.ComponentModel.Description("Ensurance of exiting from game safely and intentionally")]
```
Comment (bdach) [inline osu.Game.Tests/Visual/Menus/TestSceneSafeExiting.cs:19]:
```suggestion
    [System.ComponentModel.Description("Ensures exiting from game safely and intentionally")]
```

File: osu.Game/Screens/Menu/Exiting/SafeGameExiter.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        /// <summary>
+        /// Performs an exit from the game with safety and intention checks.
+        /// </summary>
+        /// <remarks>
+        /// If called and returned <c>true</c> at least once, all calls afterwards will early-return with <c>true</c>.
+        /// </remarks>
+        /// <param name="enforceConfirmation">Whether to enforce showing the confirmation dialog.</param>
+        /// <returns>Whether to continue exiting if it is.</returns>
```
Comment (bdach) [inline osu.Game/Screens/Menu/Exiting/SafeGameExiter.cs:57]: if it is...? I don't understand this.

File: osu.Game/Screens/Menu/Exiting/SafeGameExiter.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                onSafeExitPassed();
+                return true;
+            }
+
+            // Block exit when the notification overlay has un-cancellable progress notifications.
+            if (notificationOverlay?.HasUnCancellableWork == true)
+            {
+                notificationOverlay.Show();
+                if (!(dialogOverlay.CurrentDialog is CannotExitDialog))
+                    dialogOverlay.Push(new CannotExitDialog("There are un-cancellable work active currently.", onExitBlock));
```
Comment (bdach) [inline osu.Game/Screens/Menu/Exiting/SafeGameExiter.cs:74]: incorrect grammar (work is singular), but I'd reword overall to something like
```suggestion
                    dialogOverlay.Push(new CannotExitDialog("There is ongoing work that can't be cancelled.", onExitBlock));
```

File: osu.Game/Screens/Menu/Exiting/SafeGameExiter.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        public bool PerformSafeExit(bool enforceConfirmation)
+        {
+            if (safeExitPerformed)
+                return true;
+
+            if (dialogOverlay == null)
+            {
+                onSafeExitPassed();
+                return true;
+            }
```
Comment (bdach) [inline osu.Game/Screens/Menu/Exiting/SafeGameExiter.cs:67]: why is this branch checked before `notificationOverlay?.HasUnCancellableWork`? Requires explanation IMO.

Overall having paths return `true` before the paths that return `false` (save for checking `safeExitPerformed`) looks buggy.

**Comment by peppy** [comment]
Applications shouldn't block close ever, that seems very anti-user.

Where is this requirement coming from? I don't recall any discussion of issues mentioning this as an issue (or suggesting this as the correct direction). I think we would rather fix the cases where you think a process cannot be cancelled to be cancellable.

**Comment by frenzibyte** [comment]
> Applications shouldn't block close ever, that seems very anti-user.
> 
> Where is this requirement coming from? I don't recall any discussion of issues mentioning this as an issue (or suggesting this as the correct direction). I think we would rather fix the cases where you think a process cannot be cancelled to be cancellable.

Closing the game while it is installing a squirrel update will cause the game to never run back. (or that's what I got once while trying it)

**Comment by peppy** [comment]
Can you report that issue (including the state of the game files when this occurred) so we can fix that rather than this?

**Comment by frenzibyte** [comment]
> Can you report that issue (including the state of the game files when this occurred) so we can fix that rather than this?

Sure, will do once I get access to Windows.

**Comment by peppy** [comment]
Highly recommend asking about these kinds of things (at very least opening an issue) before spending time on implementation. To avoid wasted time.

**Comment by peppy** [comment]
Closing as incorrect direction.

Blocking user actions like exiting the application are not something we want to do.
